File: backend/src/main/java/com/urmyfood/backend/domain/repository/PostRepository.java
package com.urmyfood.backend.domain.repository;

import com.urmyfood.backend.application.dto.UpdatePostRequest;
import com.urmyfood.backend.domain.model.Post;
import com.urmyfood.backend.domain.model.PostComment;
import com.urmyfood.backend.domain.model.PostRanked;
import com.urmyfood.backend.domain.model.PostStatus;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface PostRepository {
    Post save(Post post);
    Optional<PostRanked> findRankedPostById(UUID postId, Long viewerAccountId);
    Optional<Post> findById(UUID postId);
    boolean existsById(UUID postId);
    Optional<Post> findByIdForUpdate(UUID postId);
    List<Post> findAllOrderedByCreatedAt();

    // Anchored offset for newsfeed
    List<PostRanked> findRanked(Long viewerAccountId, double w1, double w2, double w3, int page, int size, OffsetDateTime anchor);
    long countActive(OffsetDateTime anchor);

    // Anchored offset for search
    List<PostRanked> searchByKeyword(String keyword, Long viewerAccountId, int page, int size, OffsetDateTime anchor);
    long countByKeyword(String keyword, OffsetDateTime anchor);

    long likePost(UUID postId, Long accountId);
    long unlikePost(UUID postId, Long accountId);

    // Cursor-based for comments
    List<PostComment> findComments(UUID postId, OffsetDateTime cursor, int size);
    long countComments(UUID postId);
    PostComment saveComment(UUID postId, Long accountId, String content, UUID parentId);

    // Shop owner post management
    List<PostRanked> findByAuthorId(Long accountId, int page, int size);
    long countByAuthorId(Long accountId);
    void updatePost(UUID postId, Long authorId, UpdatePostRequest req);
    void updateRemainingQuantity(UUID postId, Long authorId, int remainingQuantity);
    void deletePost(UUID postId, Long authorId);
    void updatePostStatus(UUID postId, Long authorId, PostStatus status);

    // Admin methods
    List<Post> findAll(int page, int size);
    long countAll();
    void adminUpdatePostStatus(UUID postId, PostStatus status);
    void adminDeletePost(UUID postId);
}

File: frontend/app-shop/src/main/java/com/urmyfood/shop/domain/repository/PostRepository.kt
package com.urmyfood.shop.domain.repository

import com.urmyfood.shared.domain.model.Result
import com.urmyfood.shop.data.model.CreatePostRequest
import com.urmyfood.shop.data.model.UpdatePostRequest
import com.urmyfood.shop.domain.model.Comment
import com.urmyfood.shop.domain.model.Post
import okhttp3.MultipartBody

interface PostRepository {
    suspend fun getMyPosts(token: String, page: Int, size: Int): Result<List<Post>>
    suspend fun getPostById(token: String, postId: String): Result<Post>
    suspend fun createPost(token: String, request: CreatePostRequest): Result<Post>
    suspend fun updatePost(token: String, postId: String, request: UpdatePostRequest): Result<Post>
    suspend fun deletePost(token: String, postId: String): Result<Unit>
    suspend fun updatePostStatus(token: String, postId: String, status: String): Result<Post>
    suspend fun updateRemainingQuantity(token: String, postId: String, remainingQuantity: Int): Result<Post>
    suspend fun uploadPostImage(token: String, file: MultipartBody.Part): Result<String>
    suspend fun getComments(token: String, postId: String): Result<List<Comment>>
    suspend fun addComment(token: String, postId: String, content: String, parentId: String?): Result<Comment>
}
